relay/dynamic_config.go

```diff
@@ -0,0 +1,177 @@
+package relay
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	log "github.com/Sirupsen/logrus"
+	"github.com/go-yaml/yaml"
+	"github.com/operable/go-relay/relay/bus"
+	"github.com/operable/go-relay/relay/messages"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+var refreshInterval = time.Duration(5) * time.Second
+
+// DynamicConfigUpdater periodically updates bundle dynamic configurations from Cog
+type DynamicConfigUpdater struct {
+	id                string
+	configTopic       string
+	options           bus.ConnectionOptions
+	conn              bus.Connection
+	dynamicConfigRoot string
+	lastSignature     string
+	control           chan interface{}
+	refreshTimer      *time.Timer
+}
+
+// NewDynamicConfigUpdater creates a new updater
+func NewDynamicConfigUpdater(relayID string, busOpts bus.ConnectionOptions, dynamicConfigRoot string) *DynamicConfigUpdater {
+	return &DynamicConfigUpdater{
+		id:                relayID,
+		configTopic:       fmt.Sprintf("bot/relays/%s/dynconfigs", relayID),
+		options:           busOpts,
+		dynamicConfigRoot: dynamicConfigRoot,
+		control:           make(chan interface{}),
+	}
+}
+
+// Run connects the announcer to Cog and starts its main
+// loop in a goroutine
+func (dcu *DynamicConfigUpdater) Run() error {
+	dcu.options.AutoReconnect = true
+	dcu.options.EventsHandler = dcu.handleBusEvents
+	conn := &bus.MQTTConnection{}
+	if err := conn.Connect(dcu.options); err != nil {
+		return err
+	}
+	dcu.refreshTimer = time.AfterFunc(refreshInterval, dcu.refreshConfigs)
+	go func() {
+		dcu.loop()
+	}()
+	return nil
+}
+
+func (dcu *DynamicConfigUpdater) Halt() {
+	dcu.control <- 1
+}
+
+func (dcu *DynamicConfigUpdater) handleBusEvents(conn bus.Connection, event bus.Event) {
+	if event == bus.ConnectedEvent {
+		dcu.conn = conn
+		if err := dcu.conn.Subscribe(dcu.configTopic, dcu.dynConfigUpdate); err != nil {
+			log.Errorf("Failed to set up dynamic config updater subscriptions: %s.", err)
+			panic(err)
+		}
+	}
+}
+
+func (dcu *DynamicConfigUpdater) dynConfigUpdate(conn bus.Connection, topic string, payload []byte) {
+	defer dcu.refreshTimer.Reset(refreshInterval)
+	var envelope messages.DynamicConfigsResponseEnvelope
+	decoder := json.NewDecoder(bytes.NewReader(payload))
+	decoder.UseNumber()
```

Comment:
One day, we should try and pull all the JSON decoding into a module so we don't have to keep repeating this same song and dance... 😦 . I'm afraid one day we're going to add something new and forget about this little wrinkle.